src/ray/gcs/gcs_server/gcs_actor_distribution.cc

```diff
@@ -0,0 +1,227 @@
+                                          task_spec);
+
+  // TODO(Chong-Li): code path for actors that do not need a sole assignment.
+}
+
+std::shared_ptr<GcsActorWorkerAssignment>
+GcsBasedActorScheduler::AllocateNewActorWorkerAssignment(
+    const ResourceSet &required_resources, bool is_shared,
+    const TaskSpecification &task_spec) {
+  // Allocate resources from cluster.
+  auto selected_node_id = AllocateResources(required_resources);
+  if (selected_node_id.IsNil()) {
+    WarnResourceAllocationFailure(task_spec, required_resources);
+    return nullptr;
+  }
+
+  // Create a new gcs actor worker assignment.
+  auto gcs_actor_worker_assignment = std::make_shared<GcsActorWorkerAssignment>(
+      selected_node_id, required_resources, is_shared);
+
+  return gcs_actor_worker_assignment;
+}
+
+NodeID GcsBasedActorScheduler::AllocateResources(const ResourceSet &required_resources) {
+  auto selected_nodes =
+      gcs_resource_scheduler_->Schedule({required_resources}, SchedulingType::SPREAD);
+
+  if (selected_nodes.size() == 0) {
+    RAY_LOG(INFO)
+        << "Scheduling resources failed, schedule type = SchedulingType::SPREAD";
+    return NodeID::Nil();
+  }
+
+  RAY_CHECK(selected_nodes.size() == 1);
+
+  auto selected_node_id = selected_nodes[0];
+  if (!selected_node_id.IsNil()) {
+    // Acquire the resources from the selected node.
+    RAY_CHECK(
+        gcs_resource_manager_->AcquireResources(selected_node_id, required_resources));
+  }
+
+  return selected_node_id;
+}
+
+NodeID GcsBasedActorScheduler::GetHighestScoreNodeResource(
+    const ResourceSet &required_resources) const {
+  const auto &cluster_map = gcs_resource_manager_->GetClusterResources();
+
+  /// Get the highest score node
+  LeastResourceScorer scorer;
+
+  double highest_score = std::numeric_limits<double>::lowest();
+  auto highest_score_node = NodeID::Nil();
+  for (const auto &pair : cluster_map) {
+    double least_resource_val = scorer.Score(required_resources, pair.second);
+    if (least_resource_val > highest_score) {
+      highest_score = least_resource_val;
+      highest_score_node = pair.first;
+    }
+  }
+
+  return highest_score_node;
+}
+
+void GcsBasedActorScheduler::WarnResourceAllocationFailure(
+    const TaskSpecification &task_spec, const ResourceSet &required_resources) const {
+  auto scheduling_node_id = GetHighestScoreNodeResource(required_resources);
+  const SchedulingResources *scheduling_resource = nullptr;
+  auto iter = gcs_resource_manager_->GetClusterResources().find(scheduling_node_id);
+  if (iter != gcs_resource_manager_->GetClusterResources().end()) {
+    scheduling_resource = &iter->second;
+  }
+  const std::string &scheduling_resource_str =
+      scheduling_resource ? scheduling_resource->DebugString() : "None";
+  // Return nullptr if the cluster resources are not enough.
+  RAY_LOG(WARNING) << "No enough resources for creating actor "
```

Comment:
In case of lack resources to run a job, how frequently the WARNING will be printed? We'd like to avoid too bad log messages to disrupt the end-users. Could you verify that it'll be printed at a reasonable frequency? Not things like it keep printing until there is the resource.

Reply:
In my experience, it is an important message for trouble shooting and a signal for auto-scaling. In some cases, this msg has to be feedback to end-users to warn any (maybe unintentional) resource abuse.

Reply:
In our productions, we've not found this msg annoying.

Reply:
I only want to make sure this log is not keeping printing thousands of lines of logs when the resource is not enough.  I'm not against this log, if you think it's not this case.